process/interceptors/processor/sovereignHdrInterceptorProcessor.go

```diff
@@ -16,50 +21,113 @@ type ArgsSovereignHeaderInterceptorProcessor struct {
 }
 
 // NewSovereignHdrInterceptorProcessor creates a new sovereign extended header interceptor processor
-func NewSovereignHdrInterceptorProcessor(argument *ArgsSovereignHeaderInterceptorProcessor) (*sovereignHeaderInterceptorProcessor, error) {
-	if argument == nil {
-		return nil, process.ErrNilArgumentStruct
-	}
-	if check.IfNil(argument.BlockBlackList) {
-		return nil, process.ErrNilBlackListCacher
+func NewSovereignHdrInterceptorProcessor(args *ArgsSovereignHeaderInterceptorProcessor) (*sovereignHeaderInterceptorProcessor, error) {
+	err := checkSovHdrProcArgs(args)
+	if err != nil {
+		return nil, err
 	}
 
 	return &sovereignHeaderInterceptorProcessor{
-		blackList:                argument.BlockBlackList,
-		hasher:                   argument.Hasher,
-		marshaller:               argument.Marshaller,
-		IncomingHeaderSubscriber: argument.IncomingHeaderSubscriber,
+		blackList:                args.BlockBlackList,
+		hasher:                   args.Hasher,
+		marshaller:               args.Marshaller,
+		incomingHeaderSubscriber: args.IncomingHeaderSubscriber,
+		headersPool:              args.HeadersPool,
 	}, nil
 }
 
+func checkSovHdrProcArgs(args *ArgsSovereignHeaderInterceptorProcessor) error {
+	if args == nil {
+		return process.ErrNilArgumentStruct
+	}
+	if check.IfNil(args.BlockBlackList) {
+		return process.ErrNilBlackListCacher
+	}
+	if check.IfNil(args.Hasher) {
+		return errors.ErrNilHasher
+	}
+	if check.IfNil(args.Marshaller) {
+		return errors.ErrNilMarshalizer
+	}
+	if check.IfNil(args.IncomingHeaderSubscriber) {
+		return errors.ErrNilIncomingHeaderSubscriber
+	}
+	if check.IfNil(args.HeadersPool) {
+		return process.ErrNilHeadersDataPool
+	}
+
+	return nil
+}
+
 // Validate checks if the intercepted data can be processed
 func (hip *sovereignHeaderInterceptorProcessor) Validate(data process.InterceptedData, _ core.PeerID) error {
-	_, ok := data.(process.ExtendedHeaderValidatorHandler)
+	interceptedHdr, ok := data.(process.ExtendedHeaderValidatorHandler)
 	if !ok {
 		return fmt.Errorf("sovereignHeaderInterceptorProcessor.Validate error: %w", process.ErrWrongTypeAssertion)
 	}
 
+	hip.blackList.Sweep()
+	isBlackListed := hip.blackList.Has(string(interceptedHdr.Hash()))
+	if isBlackListed {
+		return process.ErrHeaderIsBlackListed
+	}
+
+	return hip.validateReceivedHeader(interceptedHdr.GetExtendedHeader(), interceptedHdr.Hash())
+}
+
+func (hip *sovereignHeaderInterceptorProcessor) validateReceivedHeader(
+	extendedHdr data.ShardHeaderExtendedHandler,
+	hash []byte,
+) error {
+	computedExtendedHeader, err := hip.incomingHeaderSubscriber.CreateExtendedHeader(extendedHdr)
+	if err != nil {
+		return err
+	}
+
+	computedHeaderHash, err := core.CalculateHash(hip.marshaller, hip.hasher, computedExtendedHeader)
+	if err != nil {
+		return err
+	}
+
+	if !bytes.Equal(computedHeaderHash, hash) {
+		return fmt.Errorf("%w, computed hash: %s, received hash: %s",
+			errors.ErrInvalidReceivedSovereignProof,
```

Comment:
It is proof or header?

Reply:
For now, it is a header. Will be a proof in the future. Changed the name, though